File: src/test/java/io/github/mschout/email/srs/provider/GuardedSRSProviderTest.java
package io.github.mschout.email.srs.provider;

import static org.junit.jupiter.api.Assertions.*;

import com.google.common.collect.ImmutableList;
import io.github.mschout.email.srs.SRS;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class GuardedSRSProviderTest {
  private SRS srs;

  @BeforeEach
  public void createSrs() {
    srs = new SRS(SRS.Type.GUARDED, "foo");
  }

  @Test
  public void forward() throws NoSuchAlgorithmException, InvalidKeyException {
    String addr = "user@domain-with-dash.com";

    String srs0 = srs.forward(addr, "foo.com");
    String srs1 = srs.forward(srs0, addr);

    assertTrue(srs0.startsWith("SRS0"), "Starts with SRS0");
    assertTrue(srs1.startsWith("SRS1"), "Starts with SRS1");
    assertEquals('=', srs1.charAt(4), "Separator is '='");

    assertTrue(srs0.matches("SRS0=\\S{4}=\\S{2}=domain-with-dash\\.com=user@foo.com"));
    assertTrue(srs1.matches("SRS1=\\S{4}=foo.com==\\S{4}=\\S{2}=domain-with-dash\\.com=user@domain-with-dash\\.com"));

    assertEquals(addr, srs.reverse(srs0), "SRS0 reverses to original address");
    assertEquals(srs0, srs.reverse(srs1), "SRS1 reverses to SRS0 address");
    assertEquals(addr, srs.reverse(srs.reverse(srs1)), "Reversal of reverse(srs1) is original address");
  }

  @Test
  public void reverse() throws NoSuchAlgorithmException, InvalidKeyException {
    String addr = "user@domain-with-dash.com";

    String alias0 = srs.forward(addr, "foo.com");
    String alias1 = srs.forward(alias0, addr);

    assertEquals(alias0, srs.reverse(alias1));
    assertEquals(addr, srs.reverse(alias0));
  }

  @Test
  public void usernames() throws NoSuchAlgorithmException, InvalidKeyException {
    List<String> addresses = ImmutableList.of(
      "user@domain-with-dash.com",
      "user-with-dash@domain.com",
      "user+with+plus@domain.com",
      "user%with!everything&everything=@domain.somewhere"
    );

    List<String> aliases = ImmutableList.of("user1@tld1.com", "user2@tld2.com");

    for (String email : addresses) {
      String srs0 = srs.forward(email, aliases.get(0));
      assertEquals(email, srs.reverse(srs0), "Reverses to original address");

      // idempotent on srs0addr
      String srs1 = srs.forward(srs0, aliases.get(1));
      assertEquals(srs0, srs.reverse(srs1), "SRS1 reverses to SRS0 address");

      // idempotent from same domain.
      assertEquals(srs0, srs.forward(srs0, aliases.get(0)), "Rewruite to same domain");
    }
  }

  @Test
  public void invalidHash() throws NoSuchAlgorithmException, InvalidKeyException {
    String srs0 = srs.forward("user@domain.com", "example.com");

    // Replace hash with XXXX
    final String invalidsrs = "SRS0=XXXX" + srs0.substring(srs0.indexOf('=', 8));

    Exception exception = assertThrows(IllegalArgumentException.class, () -> srs.reverse(invalidsrs));

    assertTrue(exception.getMessage().contains("Invalid address hash: XXXX"));
  }

  @Test
  public void caseSensitivity() throws NoSuchAlgorithmException, InvalidKeyException {
    List<String> addresses = ImmutableList.of(
      "User@domain-with-dash.com",
      "User-with-dash@domain.com",
      "User+with+plus@domain.com",
      "User%with!everything&everything=@domain.somewhere"
    );

    String alias0 = "user0@tld0.com";
    String alias1 = "user1@tld1.com";

    for (String email : addresses) {
      String srs0 = srs.forward(email, alias0).toLowerCase();
      String srsRev = srs.reverse(srs0);
      assertTrue(srsRev.equalsIgnoreCase(email), String.format("%s reverses to %s case insensitively", srs0, email));

      String srs1 = srs.forward(srs0, alias1).toLowerCase();
      srsRev = srs.reverse(srs1);
      assertTrue(srsRev.equalsIgnoreCase(srs0));
    }
  }
}

File: src/test/kotlin/io/github/mschout/email/srs/provider/GuardedSRSProviderTest.kt
/*
 * Copyright 2026 Michael Schout
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.mschout.email.srs.provider

import io.github.mschout.email.srs.SRS
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldMatch
import io.kotest.matchers.string.shouldStartWith

class GuardedSRSProviderTest :
    FunSpec({
      lateinit var srs: SRS

      beforeEach { srs = SRS(SRS.Type.GUARDED, "foo") }

      test("forward") {
        val addr = "user@domain-with-dash.com"

        val srs0 = srs.forward(addr, "foo.com")
        val srs1 = srs.forward(srs0, addr)

        srs0 shouldStartWith "SRS0"
        srs1 shouldStartWith "SRS1"
        srs1[4] shouldBe '='

        srs0 shouldMatch Regex("SRS0=\\S{4}=\\S{2}=domain-with-dash\\.com=user@foo.com")
        srs1 shouldMatch
            Regex(
                "SRS1=\\S{4}=foo.com==\\S{4}=\\S{2}=domain-with-dash\\.com=user@domain-with-dash\\.com"
            )

        srs.reverse(srs0) shouldBe addr
        srs.reverse(srs1) shouldBe srs0
        srs.reverse(srs.reverse(srs1)) shouldBe addr
      }

      test("reverse") {
        val addr = "user@domain-with-dash.com"

        val alias0 = srs.forward(addr, "foo.com")
        val alias1 = srs.forward(alias0, addr)

        srs.reverse(alias1) shouldBe alias0
        srs.reverse(alias0) shouldBe addr
      }

      test("usernames") {
        val addresses =
            listOf(
                "user@domain-with-dash.com",
                "user-with-dash@domain.com",
                "user+with+plus@domain.com",
                "user%with!everything&everything=@domain.somewhere",
            )

        val aliases = listOf("user1@tld1.com", "user2@tld2.com")

        for (email in addresses) {
          val srs0 = srs.forward(email, aliases[0])
          srs.reverse(srs0) shouldBe email

          val srs1 = srs.forward(srs0, aliases[1])
          srs.reverse(srs1) shouldBe srs0

          // idempotent from same domain
          srs.forward(srs0, aliases[0]) shouldBe srs0
        }
      }

      test("invalidHash") {
        val srs0 = srs.forward("user@domain.com", "example.com")

        val invalidsrs = "SRS0=XXXX" + srs0.substring(srs0.indexOf('=', 8))

        val exception = shouldThrow<IllegalArgumentException> { srs.reverse(invalidsrs) }
        exception.message shouldContain "Invalid address hash: XXXX"
      }

      test("caseSensitivity") {
        val addresses =
            listOf(
                "User@domain-with-dash.com",
                "User-with-dash@domain.com",
                "User+with+plus@domain.com",
                "User%with!everything&everything=@domain.somewhere",
            )

        val alias0 = "user0@tld0.com"
        val alias1 = "user1@tld1.com"

        for (email in addresses) {
          val srs0 = srs.forward(email, alias0).lowercase()
          val srsRev = srs.reverse(srs0)
          srsRev.equals(email, ignoreCase = true).shouldBeTrue()

          val srs1 = srs.forward(srs0, alias1).lowercase()
          val srs1Rev = srs.reverse(srs1)
          srs1Rev.equals(srs0, ignoreCase = true).shouldBeTrue()
        }
      }
    })
